plutus-core/src/Language/PlutusCore/Evaluation/Machine/Ck.hs

```diff
@@ -337,36 +369,63 @@ applyBuiltin
     => Context uni fun
     -> fun
     -> [CkValue uni fun]
-    -> CkM uni fun (Term TyName Name uni fun ())
+    -> CkM uni fun s (Term TyName Name uni fun ())
 applyBuiltin stack bn args = do
-    let ckValueToTermInError = hoist $ first $ mapCauseInMachineException ckValueToTerm
+    let dischargeError = hoist $ withExceptT $ mapCauseInMachineException ckValueToTerm
     BuiltinRuntime sch _ f exF <- asksM $ lookupBuiltin bn . ckEnvRuntime
-    result <- ckValueToTermInError $ applyTypeSchemed bn sch f exF args
+    result <- dischargeError $ applyTypeSchemed bn sch f exF args
     stack <| result
 
--- | Evaluate a term using the CK machine. May throw a 'CkEvaluationException'.
+runCk
```

Comment:
That wouldn't work as the types are different:

- `evaluateCek` returns a `(Either (CekEvaluationException uni fun) (Term Name uni fun ()), [String])`
- `evaluateCekNoEmit` returns a `Either (CekEvaluationException uni fun) (Term Name uni fun ())` (note the absence of `String`)
- `unsafeEvaluateCek` returns a `(EvaluationResult (Term Name uni fun ()), [String])` (note the absence of `Either`)
- `unsafeEvaluateCekNoEmit` returns a `EvaluationResult (Term Name uni fun ())` (note the absence of `Either` and `[String]`)

Deferring this result type juggling to use-site would only make things more obscure.

But maybe we can have this proliferation only in `runCek*` functions and derive `evaluateCek*` ones using a single combinator, I'll think about it.

Reply:
I spent an hour staring at that API and while I think it's possible to do

> maybe we can have this proliferation only in runCek* functions and derive evaluateCek* ones using a single combinator

by having something like

```haskell
nobudgeting :: (ExBudgetMode -> a -> (b, CekExBudgetState fun)) -> a -> b
nobudgeting f = fst . f Counting
```

it doesn't seem like that would add any clarity, especially if it appears at use-site rather than internally. And even internally it doesn't seem to make things any more clear as we only use this pattern once in

```haskell
evaluateCekNoEmit runtime = fst . runCekNoEmit runtime Counting
```

and hence it's not worth introducing a general definition.

Perhaps we should revise this later if the API changes, but for now I don't see what we can do to improve it. So I'm resolving this.